File: src/se/clau/intellijlux/LuxParserDefinition.java
package se.clau.intellijlux;

import com.intellij.lang.*;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.tree.*;
import se.clau.intellijlux.parser.LuxParser;
import se.clau.intellijlux.psi.*;
import org.jetbrains.annotations.NotNull;

public class LuxParserDefinition implements ParserDefinition {
  public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
  public static final TokenSet COMMENTS = TokenSet.create(LuxTypes.COMMENT);

  public static final IFileElementType FILE = new IFileElementType(LuxLanguage.INSTANCE);

  @NotNull
  @Override
  public Lexer createLexer(Project project) {
    return new LuxLexerAdapter();
  }

  @NotNull
  public TokenSet getWhitespaceTokens() {
    return WHITE_SPACES;
  }

  @NotNull
  public TokenSet getCommentTokens() {
    return COMMENTS;
  }

  @NotNull
  public TokenSet getStringLiteralElements() {
    return TokenSet.EMPTY;
  }

  @NotNull
  public PsiParser createParser(final Project project) {
    return new LuxParser();
  }

  @Override
  public IFileElementType getFileNodeType() {
    return FILE;
  }

  public PsiFile createFile(FileViewProvider viewProvider) {
    return new LuxFile(viewProvider);
  }

  public SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
    return SpaceRequirements.MAY;
  }

  @NotNull
  public PsiElement createElement(ASTNode node) {
    return LuxTypes.Factory.createElement(node);
  }
}

File: src/se/clau/intellijlux/LuxParserDefinition.kt
package se.clau.intellijlux

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.ParserDefinition.SpaceRequirements
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import se.clau.intellijlux.gen.parser.LuxParser
import se.clau.intellijlux.gen.psi.LuxTypes

class LuxParserDefinition : ParserDefinition {
  override fun createLexer(project: Project): Lexer = LuxLexerAdapter()
  override fun getWhitespaceTokens(): TokenSet = TokenSet.create(TokenType.WHITE_SPACE)
  override fun getCommentTokens(): TokenSet = TokenSet.create(LuxTypes.COMMENT)
  override fun getStringLiteralElements(): TokenSet = TokenSet.EMPTY
  override fun createParser(project: Project): PsiParser = LuxParser()

  companion object {
    val FILE = IFileElementType(LuxLanguage)
  }

  override fun getFileNodeType(): IFileElementType = FILE

  override fun createFile(viewProvider: FileViewProvider): PsiFile = LuxFile(viewProvider)
  override fun spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode): SpaceRequirements =
    SpaceRequirements.MAY

  override fun createElement(node: ASTNode): PsiElement = LuxTypes.Factory.createElement(node)
}
